File: voice-engine/native/apple/WhisperKitSTT/Sources/WhisperKitSTT/EventEmitter.swift
//
//  EventEmitter.swift
//
//  Shared JSONL event emission. One line per event on stdout. Mirrors the
//  TranscriptEvent shape used by the Python lab and the AppleSTT sibling
//  CLI; the lab's strategy wrappers all parse the same field set.
//
//  Shape:
//    {
//      "type": "partial" | "final",
//      "text": "...",
//      "timestamp_ms": Int,                  // audio-relative
//      "latency_ms_from_audio_start": Int,   // wall-clock from audio start
//      "confidence": Double?,                // exp(avg_logprob) clamped 0..1
//      "engine_metadata": { ... }            // free-form
//    }
//
//  We hand-encode the JSON line to keep ordering stable, keys consistent,
//  and avoid pulling in extra dependencies. Stdout is line-buffered via
//  setvbuf so consumers see partials as they happen.

import Foundation

#if canImport(Darwin)
import Darwin
#endif

enum EventType: String {
    case partial
    case final
}

struct TranscriptEventOut {
    let type: EventType
    let text: String
    let timestampMs: Int
    let latencyMsFromAudioStart: Int
    let confidence: Double?
    let engineMetadata: [String: Any]
}

final class EventEmitter {
    /// Wall-clock origin used to compute `latency_ms_from_audio_start`.
    /// The runner sets this once, right before it kicks off transcription.
    private(set) var audioStart: Date

    /// Serialize emit() calls. WhisperKit fires callbacks from arbitrary
    /// Tasks, so emissions can race; an NSLock guarantees one JSONL line
    /// at a time on stdout.
    private let lock = NSLock()

    init(audioStart: Date = Date()) {
        self.audioStart = audioStart
        // Best-effort line buffering on stdout so Python sees partials live.
        setvbuf(stdout, nil, _IOLBF, 0)
    }

    func resetAudioStart(_ date: Date = Date()) {
        self.audioStart = date
    }

    func emit(_ event: TranscriptEventOut) {
        let payload: [String: Any?] = [
            "type": event.type.rawValue,
            "text": event.text,
            "timestamp_ms": event.timestampMs,
            "latency_ms_from_audio_start": event.latencyMsFromAudioStart,
            "confidence": event.confidence,
            "engine_metadata": event.engineMetadata
        ]
        // Preserve `confidence: null` rather than dropping the key.
        let cleaned: [String: Any] = payload.compactMapValues { value in
            if value == nil {
                return NSNull()
            }
            return value
        }
        do {
            let data = try JSONSerialization.data(
                withJSONObject: cleaned,
                options: []
            )
            if let line = String(data: data, encoding: .utf8) {
                lock.lock()
                defer { lock.unlock() }
                FileHandle.standardOutput.write(Data((line + "\n").utf8))
            }
        } catch {
            FileHandle.standardError.write(
                Data("error: failed to encode event: \(error)\n".utf8)
            )
        }
    }

    /// Convenience to compute the wall-clock latency from audio start.
    func latencyMsSinceAudioStart() -> Int {
        return Int(Date().timeIntervalSince(audioStart) * 1000.0)
    }
}

/// Emit a structured error to stderr and exit non-zero. Python wrapper
/// surfaces these as TranscriptionError with a fix-it hint.
func fail(_ message: String, code: Int32 = 1) -> Never {
    FileHandle.standardError.write(Data("WhisperKitSTT error: \(message)\n".utf8))
    exit(code)
}

File: voice-engine/native/android/SherpaOnnxSTT/src/main/kotlin/com/trika/sherpaonnxstt/EventEmitter.kt
package com.trika.sherpaonnxstt

import java.io.PrintStream

/**
 * JSONL event emitter — mirrors the Swift EventEmitter used by WhisperKitSTT.
 *
 * Shape (one line per event):
 *   {
 *     "type": "partial" | "final",
 *     "text": "...",
 *     "timestamp_ms": Int,                  // audio-relative (ms)
 *     "latency_ms_from_audio_start": Int,   // wall-clock from audio start (ms)
 *     "confidence": Double | null,
 *     "engine_metadata": { ... }
 *   }
 *
 * Stdout is flushed after every event so the Python subprocess driver sees
 * lines as they arrive (equivalent to Swift's _IOLBF line-buffering).
 */
object EventEmitter {

    /** Wall-clock origin set once by the runner just before transcription. */
    @Volatile
    private var audioStartMs: Long = System.currentTimeMillis()

    private val out: PrintStream = System.out

    fun resetAudioStart() {
        audioStartMs = System.currentTimeMillis()
    }

    fun latencyMsSinceAudioStart(): Int =
        (System.currentTimeMillis() - audioStartMs).toInt()

    /**
     * Emit a `partial` event. `confidence` may be null (Whisper offline gives
     * no per-token probability for the simulated partial).
     */
    fun emitPartial(
        text: String,
        timestampMs: Int,
        latencyMs: Int,
        confidence: Double?,
        engineMetadata: Map<String, Any?>,
    ) = emitEvent(
        type = "partial",
        text = text,
        timestampMs = timestampMs,
        latencyMs = latencyMs,
        confidence = confidence,
        engineMetadata = engineMetadata,
    )

    /**
     * Emit a `final` event.
     */
    fun emitFinal(
        text: String,
        timestampMs: Int,
        latencyMs: Int,
        confidence: Double?,
        engineMetadata: Map<String, Any?>,
    ) = emitEvent(
        type = "final",
        text = text,
        timestampMs = timestampMs,
        latencyMs = latencyMs,
        confidence = confidence,
        engineMetadata = engineMetadata,
    )

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    @Synchronized
    private fun emitEvent(
        type: String,
        text: String,
        timestampMs: Int,
        latencyMs: Int,
        confidence: Double?,
        engineMetadata: Map<String, Any?>,
    ) {
        val sb = StringBuilder()
        sb.append("{")
        sb.appendJsonString("type", type); sb.append(",")
        sb.appendJsonString("text", text); sb.append(",")
        sb.appendJsonInt("timestamp_ms", timestampMs); sb.append(",")
        sb.appendJsonInt("latency_ms_from_audio_start", latencyMs); sb.append(",")
        if (confidence == null) {
            sb.append("\"confidence\":null")
        } else {
            sb.append("\"confidence\":${confidence}")
        }
        sb.append(",")
        sb.append("\"engine_metadata\":{")
        val metaEntries = engineMetadata.entries.toList()
        metaEntries.forEachIndexed { i, (k, v) ->
            sb.appendJsonValue(k, v)
            if (i < metaEntries.size - 1) sb.append(",")
        }
        sb.append("}")
        sb.append("}")

        out.println(sb.toString())
        out.flush()
    }

    // Minimal JSON helpers — avoids a heavy JSON library dep at runtime.

    private fun StringBuilder.appendJsonString(key: String, value: String) {
        append("\"").append(key).append("\":\"").append(escapeJson(value)).append("\"")
    }

    private fun StringBuilder.appendJsonInt(key: String, value: Int) {
        append("\"").append(key).append("\":").append(value)
    }

    private fun StringBuilder.appendJsonValue(key: String, value: Any?) {
        append("\"").append(escapeJson(key)).append("\":")
        when (value) {
            null -> append("null")
            is Boolean -> append(value.toString())
            is Int -> append(value.toString())
            is Long -> append(value.toString())
            is Float -> append(value.toString())
            is Double -> append(value.toString())
            is String -> append("\"").append(escapeJson(value)).append("\"")
            else -> append("\"").append(escapeJson(value.toString())).append("\"")
        }
    }

    private fun escapeJson(s: String): String = s
        .replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
}
